File: 2023/19/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n\n");

const workflowRegex = /(.+){(.+)}/;
const ruleRegex = /(\w+)([><])(\d+):(\w+)/;

class Workflow {
  constructor(line) {
    const regexMatch = line.match(workflowRegex);
    this.name = regexMatch[1];
    this.rules = regexMatch[2].split(",").map((rule) => {
      const ruleMatch = rule.match(ruleRegex);
      if (ruleMatch) {
        const category = ruleMatch[1];
        const num = Number(ruleMatch[3]);
        const sign = ruleMatch[2];
        const test =
          sign === ">"
            ? (part) => part[category] > num
            : (part) => part[category] < num;
        return {
          category,
          num,
          rule,
          sign,
          test,
          next: ruleMatch[4],
        };
      } else {
        return {
          category: "?",
          num: 0,
          sign: "=",
          rule,
          test: (part) => true,
          next: rule,
        };
      }
    });
  }

  evaluatePart(part) {
    for (const rule of this.rules) {
      if (rule.test(part)) {
        return rule.next;
      }
    }
  }
}

const workflowMap = new Map();
input[0].split("\n").forEach((line) => {
  const workflow = new Workflow(line);
  workflowMap.set(workflow.name, workflow);
});
const ratings = input[1].split("\n").map((line) => {
  return line
    .slice(1, -1)
    .split(",")
    .reduce((acc, part) => {
      const [category, num] = part.split("=");
      acc[category] = Number(num);
      return acc;
    }, {});
});

let part1 = 0;
for (const rating of ratings) {
  let curr = "in";
  while (curr !== "A" && curr !== "R") {
    curr = workflowMap.get(curr).evaluatePart(rating);
  }
  if (curr === "A") {
    part1 += Object.values(rating).reduce((acc, val) => acc + val, 0);
  }
}
console.log(part1);

function reverseRule(rule) {
  const newRule = { ...rule };
  newRule.sign = rule.sign === ">" ? "<" : ">";
  newRule.rule = rule.rule.replace(rule.sign, newRule.sign);
  newRule.num = rule.num + (rule.sign === ">" ? 1 : -1);
  newRule.test =
    rule.sign === ">"
      ? (part) => part[rule.category] < newRule.num
      : (part) => part[rule.category] > newRule.num;
  return newRule;
}

const queue = [{ rules: [], currentWorkflow: workflowMap.get("in") }];
const combinations = [];
while (queue.length) {
  const { rules, currentWorkflow } = queue.pop();
  // for each rule, reverse all rules before it and add current rule
  for (let i = 0; i < currentWorkflow.rules.length; i++) {
    const currentRules = rules.slice();
    for (let j = 0; j < i; j++) {
      currentRules.push(reverseRule(currentWorkflow.rules[j]));
    }
    currentRules.push(currentWorkflow.rules[i]);
    // if next is "A", add to combinations
    // if next is "R", do nothing (discard)
    // otherwise, add to queue
    if (currentWorkflow.rules[i].next === "A") {
      combinations.push(currentRules);
    } else if (currentWorkflow.rules[i].next !== "R") {
      queue.push({
        rules: currentRules,
        currentWorkflow: workflowMap.get(currentWorkflow.rules[i].next),
      });
    }
  }
}

let part2 = 0;
for (const combo of combinations) {
  const approved = { x: [1, 4000], m: [1, 4000], a: [1, 4000], s: [1, 4000] };
  for (const rule of combo) {
    if (rule.sign === "=") continue;
    if (rule.sign === ">") {
      approved[rule.category][0] = Math.max(
        approved[rule.category][0],
        rule.num + 1
      );
    } else {
      approved[rule.category][1] = Math.min(
        approved[rule.category][1],
        rule.num - 1
      );
    }
  }
  if (!Object.values(approved).every(([min, max]) => min <= max)) continue;

  part2 += Object.values(approved).reduce(
    (acc, [min, max]) => acc * (max - min + 1),
    1
  );
}

console.log(part2);

File: 2023/19/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"slices"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	workflowMap, parts := parseInput(strings.Split(string(data), "\n\n"))
	fmt.Println(part1(workflowMap, parts))
	fmt.Println(part2(workflowMap))
}

var (
	workflowRegex = regexp.MustCompile(`^(\w+){(.*)}$`)
	ruleRegex     = regexp.MustCompile(`^([xmas])([<>])(\d+):(\w+)$`)
	partRegex     = regexp.MustCompile(`^{x=(\d+),m=(\d+),a=(\d+),s=(\d+)}$`)
)

type Rule struct {
	category byte
	comp     byte
	n        int
	dst      string
}

func NewRule(str string) *Rule {
	match := ruleRegex.FindStringSubmatch(str)
	if match == nil {
		return &Rule{'x', '>', 0, str}
	}
	n, err := strconv.Atoi(match[3])
	if err != nil {
		panic(err)
	}
	return &Rule{match[1][0], match[2][0], n, match[4]}
}

func (r *Rule) AssessPart(p Part) bool {
	if r.comp == '<' {
		return p[r.category] < r.n
	}
	return p[r.category] > r.n
}

func (r *Rule) Reverse() *Rule {
	var comp byte = '<'
	if comp == r.comp {
		comp = '>'
	}
	n := r.n
	if r.comp == '>' {
		n++
	} else {
		n--
	}
	return &Rule{r.category, comp, n, r.dst}
}

type Workflow struct {
	name  string
	rules []*Rule
}

func NewWorkflow(str string) *Workflow {
	match := workflowRegex.FindStringSubmatch(str)
	if match == nil {
		panic(fmt.Sprintf("Error matching workflow regex with: %s", str))
	}
	name := match[1]
	ruleParts := strings.Split(match[2], ",")
	rules := make([]*Rule, len(ruleParts))
	for j, rulePart := range ruleParts {
		rules[j] = NewRule(rulePart)
	}
	return &Workflow{name, rules}
}

func (w *Workflow) AssessPart(p Part) string {
	for _, rule := range w.rules {
		if rule.AssessPart(p) {
			return rule.dst
		}
	}
	return ""
}

type Part map[byte]int

func NewPart(str string) Part {
	match := partRegex.FindStringSubmatch(str)
	if match == nil {
		panic(fmt.Sprintf("Error matching part regex with: %s", str))
	}
	x, err := strconv.Atoi(match[1])
	if err != nil {
		panic(err)
	}
	m, err := strconv.Atoi(match[2])
	if err != nil {
		panic(err)
	}
	a, err := strconv.Atoi(match[3])
	if err != nil {
		panic(err)
	}
	s, err := strconv.Atoi(match[4])
	if err != nil {
		panic(err)
	}
	return Part{'x': x, 'm': m, 'a': a, 's': s}
}

func (p Part) Rating() int {
	return p['x'] + p['m'] + p['a'] + p['s']
}

func (p Part) Assess(workflowMap map[string]*Workflow) bool {
	curr := "in"
	for curr != "A" && curr != "R" {
		curr = workflowMap[curr].AssessPart(p)
	}
	return curr == "A"
}

func parseInput(data []string) (map[string]*Workflow, []Part) {
	workflowLines := strings.Split(data[0], "\n")
	workflowMap := make(map[string]*Workflow, len(workflowLines))
	partLines := strings.Split(data[1], "\n")
	parts := make([]Part, len(partLines))

	for _, line := range workflowLines {
		workflow := NewWorkflow(line)
		workflowMap[workflow.name] = workflow
	}

	for i, line := range partLines {
		parts[i] = NewPart(line)
	}

	return workflowMap, parts
}

func part1(workflowMap map[string]*Workflow, parts []Part) int {
	res := 0
	for _, part := range parts {
		if part.Assess(workflowMap) {
			res += part.Rating()
		}
	}
	return res
}

type QueueEntry struct {
	rules    []*Rule
	workflow *Workflow
}

func part2(workflowMap map[string]*Workflow) int {
	queue := []QueueEntry{{[]*Rule{}, workflowMap["in"]}}
	var combinations [][]*Rule
	for len(queue) > 0 {
		rules := queue[0].rules
		workflow := queue[0].workflow
		queue = queue[1:]
		for i := range workflow.rules {
			currentRules := slices.Clone(rules)
			for j := range i {
				currentRules = append(currentRules, workflow.rules[j].Reverse())
			}
			currentRules = append(currentRules, workflow.rules[i])

			if workflow.rules[i].dst == "A" {
				combinations = append(combinations, currentRules)
			} else if workflow.rules[i].dst != "R" {
				queue = append(queue, QueueEntry{currentRules, workflowMap[workflow.rules[i].dst]})
			}
		}
	}

	res := 0
	for _, combo := range combinations {
		approved := map[byte][]int{'x': {1, 4000}, 'm': {1, 4000}, 'a': {1, 4000}, 's': {1, 4000}}
		for _, rule := range combo {
			if rule.comp == '>' {
				if approved[rule.category][0] < rule.n+1 {
					approved[rule.category][0] = rule.n + 1
				}
			} else {
				if approved[rule.category][1] > rule.n-1 {
					approved[rule.category][1] = rule.n - 1
				}
			}
		}
		count := 1
		for _, bounds := range approved {
			count *= bounds[1] - bounds[0] + 1
		}
		res += count
	}
	return res
}
